# Challenge2 - MNIST Handwritten Digit Recognition | #100MLProjects | #Laxmena

100MLProjects Pre Project Update: [Blog: Challenge-2 MNIST Digit Recognition](https://medium.com/@laxmena/project2-of-100mlprojetcs-classification-mnist-digit-recognition-d9208856f1f2)

- LinkedIn: https://www.linkedin.com/in/lakshmanan-meiyappan/
- Github: https://github.com/laxmena/


## SVM Classifier


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.metrics import classification_report
import pickle 
import gzip
import random
import joblib

In [2]:
def load_mnist(filename, type, n_datapoints):
    # MNIST Images have 28*28 pixels dimension
    image_size = 28
    f = gzip.open(filename)
    
    if(type == 'image'):
        f.read(16)    # Skip Non-Image information
        buf = f.read(n_datapoints * image_size * image_size)
        data = np.frombuffer(buf, dtype=np.uint8).astype(np.float32)
        data = data.reshape(n_datapoints, image_size*image_size)
    elif(type == 'label'):
        f.read(8) # Skip Inessential information
        buf = f.read(n_datapoints)
        data = np.frombuffer(buf, dtype=np.uint8).astype(np.int64)
        data = data.reshape(n_datapoints, 1)
    return data

Training Set comprises of 60K images and labels.
Test set comprises of 10K images and labels.

In [3]:
# Training Dataset
train_size = 60000
test_size = 10000
dirpath = ''
X = load_mnist(dirpath + 'train-images-idx3-ubyte.gz', 'image', train_size)
y = load_mnist(dirpath + 'train-labels-idx1-ubyte.gz', 'label', train_size)
X_test = load_mnist(dirpath + 't10k-images-idx3-ubyte.gz', 'image', test_size)
y_test = load_mnist(dirpath + 't10k-labels-idx1-ubyte.gz', 'label', test_size)

## Build Classifier

sklearn(scikit learn) is a Machine Learning library that contains most of the popular Machine Learning Models implementations.

Creating and training a SVM Classifier.

In [4]:
classifier = SVC()
classifier.fit(X, y)
# classifier.fit(X.reshape(X.shape[0], 28*28), y)
joblib.dump(classifier,'svc_m.gz',compress=('gzip',3))

C:\Users\manib\miniconda3\envs\menv\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


['svc_m.gz']

In [ ]:
classifier.score(X_test, y_test)
# classifier.score(X_test.reshape(X_test.shape[0], 28*28), y_test)

0.9792

Accuracy of the classifier is 97.92%

## Saving the Model

The trained SVC Model is saved, so that the model can be reconstructed and used in different python modules or applications.

Joblib is a library, that allows us to save a copy of runtime data in a file, and load it back whenever we need it.

- joblib.dump - Save the runtime data in a file
- joblib.load - Load the saved data from file to the runtime

Other option is to use pickle. But using pickle can result in very large file size. KNN model trained on 60K MNIST 28*28 images can be around 420MB.

Joblib offers compression. The same KNN model can weigh only 4MB if saved using compression enabled joblib.

In [ ]:
joblib.dump(classifier,'svc_m.gz',compress=('gzip',3))

['svc_mnist_60k.gz']